src/firewall/functions.py

```diff
@@ -427,6 +428,215 @@ def mac_unparse(mac):
+                    # we can use as a split.
+                    for i in range(len(port)):
+                        if port[i] != "-":
+                            continue
+                        try:
+                            p1 = parse_one(port[:i])
+                            p2 = parse_one(port[i + 1 :])
+                        except ValueError:
+                            continue
+                        if port_id1 is not None:
+                            # no unique match. We fail.
+                            raise ValueError("port name is ambiguous")
+                        port_id1, port_name1 = p1
+                        port_id2, port_name2 = p2
+
+        else:
+            # Usually, our parse just accepts strings. However, also accept
+            # already pre-parsed input.
+            if isinstance(port, int):
+                port_id1 = port
+            elif allow_range and (isinstance(port, tuple) or isinstance(port, list)):
+                if len(port) == 1:
+                    (port_id1,) = port
+                elif len(port) == 2:
+                    (port_id1, port_id2) = port
+
+        if port_id1 is None:
+            if parse_one_error is not None:
+                raise ValueError(str(parse_one_error))
+            raise ValueError("not a valid port")
+
+        if not check_range(port_id1) or (
+            port_id2 is not None and not check_range(port_id2)
+        ):
+            raise ValueError("port out of range")
+
+        port_id1, port_name1, port_id2, port_name2 = port_swap(
+            port_id1, port_name1, port_id2, port_name2
+        )
+
+    return proto, port_id1, port_name1, port_id2, port_name2
+
+
+def port_unparse(proto, port_id1, port_name1, port_id2, port_name2, *, delimiter=None):
+
+    if proto is not None:
+        if port_id1 is None:
+            # special case for icmp/icmpv6/ipv6-icmp. There is only a port_name.
+            p = port_name1
+        else:
+            p = port_unparse(
+                None, port_id1, port_name1, port_id2, port_name2, delimiter=delimiter
+            )
+        return f"{proto}:{p}"
+
+    p1 = port_name1 or str(port_id1)
+    p2 = None
+    if port_id2 is not None:
+        p2 = port_name2 or str(port_id2)
+
+    if delimiter is not None:
+        if p2 is None:
+            return p1
+        return f"{p1}{delimiter}{p2}"
+
+    # We want to unparse something, that can be parsed back. Since we use
+    # '-' as delimiter, and '-' can be part of the names, it's not entirely
+    # clear that we always can.
+    #
+    # If any of the names contain a '-', join them with a space. Our names
+    # cannot contain spaces, but the parser strips spaces around the "-"
+    if p2 is None:
+        return p1
+    if "-" not in p1 and "-" not in p2:
+        return f"{p1}-{p2}"
+    return f"{p1} - {p2}"
+
+
+port_check = _parse_check_fcn(port_parse)
+
+port_norm = _parse_norm_fcn(port_parse, port_unparse)
```

Comment:
The `EntryType` class exists as a kind of Enum for identifying the type of a string entry.

When we have a string, it might be of type `EntryTypeAddr` or `EntryTypeMask`. You can find out, via `firewall.functions.EntryType.parse()`. It parses the string, gives you the detected entry-type and the decomposed data (`detail`).

Afterwards, you can treat the result generically. Of course, you can ask `if entrytype is firewall.functions.EntryTypeAddr:`. But you can also stringify the details again by calling generically `entrytype.unparse()`.

---

If you know/expect to deal with an port string, there is no reason to bother with an `EntryTypePort`. Then you have `port_parse()`, `port_unparse()` (and based on those two, there is also `port_check()` and `port_norm()`). Granted, these two are currently not used. But that's because "port" is a relatively seldom used type. The respective `ipaddr_check()` and `ipaddr_norm()` are in use. `port_check()` and `port_norm()` exists for completeness, because they go together with the parse/unparse methods and because other types have them too (`ipaddr_check()`).